SalesforceSDK/Analytics/Model/InstrumentationEventBuilder.cs

```diff
@@ -0,0 +1,197 @@
+        }
+
+        public InstrumentationEvent.EventType EventType
+        {
+            get { return _eventType; }
+            set { _eventType = value; }
+        }
+
+        public InstrumentationEvent.ErrorType ErrorType
+        {
+            get { return _errorType; }
+            set { _errorType = value; }
+        }
+
+        public string SenderParentId
+        {
+            get { return _senderParentId; }
+            set { _senderParentId = value; }
+        }
+
+        public long SessionStartTime
+        {
+            get { return _sessionStartTime; }
+            set { _sessionStartTime = value; }
+        }
+
+        public JObject Page
+        {
+            get { return _page; }
+            set { _page = value; }
+        }
+
+        public JObject PreviousPage
+        {
+            get { return _previousPage; }
+            set { _previousPage = value; }
+        }
+
+        public JObject Marks
+        {
+            get { return _marks; }
+            set { _marks = value; }
+        }
+
+        public InstrumentationEvent BuildEvent()
+        {
+            string eventId = Guid.NewGuid().ToString();
+            if (string.IsNullOrEmpty(Name))
+            {
+                throw new EventBuilderException("Mandatory field 'name' not set!");
+            }
+            var deviceAppAttributes = AnalyticsManager.GetDeviceAppAttributes();
+            if (deviceAppAttributes == null)
+            {
+                throw new EventBuilderException("Mandatory field 'device app attributes' not set!");
+            }
+            if (SchemaType == InstrumentationEvent.SchemaType.LightningInteraction
+                || SchemaType == InstrumentationEvent.SchemaType.LightningPerformance)
+            {
+                throw new EventBuilderException("Mandatory field 'event type' not set!");
+            }
+            if (SchemaType != InstrumentationEvent.SchemaType.LightningPerformance && Page == null)
+            {
+                throw new EventBuilderException("Mandatory field 'page' not set!");
+            }
+
+            var sequenceId = AnalyticsManager.GetGlobalSequenceId() + 1;
+            AnalyticsManager.SetGlobalSequenceId(sequenceId);
+
+            // Defaults to current time if not explicitly set.
+            long curTime = DateTime.Now.Ticks;
+            StartTime = StartTime == 0 ? curTime : StartTime;
+            SessionStartTime = SessionStartTime == 0 ? curTime : SessionStartTime;
+            return new InstrumentationEvent(eventId, StartTime, EndTime, Name, Attributes, SessionId,
+                    sequenceId, SenderId, SenderContext, SchemaType, EventType, ErrorType,
+                    deviceAppAttributes, GetConnectionType(), SenderParentId, SessionStartTime, Page,
+                    PreviousPage, Marks);
+        }
+
+        public string GetConnectionType()
+        {
+            return ApplicationService.GetConnectionType();
```

Comment:
Wow - it's that simple? What does this typically return? Does it return subtype too?